File: Python3/599. Minimum Index Sum of Two Lists.py
class Solution:
    def findRestaurant(self, list1, list2):
        """
        :type list1: List[str]
        :type list2: List[str]
        :rtype: List[str]
        """
        indexsumDict = {} # 128ms
        common = list(set(list1) & set(list2)) 
        for name in common:
            count = list1.index(name) + list2.index(name)
            if count not in indexsumDict:
                indexsumDict[count] = [name]
            else:
                indexsumDict[count].append(name)
        return indexsumDict[min(indexsumDict.keys())]

File: Golang/599. Minimum Index Sum of Two Lists.go
func findRestaurant(list1 []string, list2 []string) []string { // 52-56ms
    indexsumDict := make(map[string]int, len(list1)+len(list2))
    for index, name := range list1 {
        indexsumDict[name] = index + 1000
    }
    min := 4000
    for index, name := range list2 {
        if indexsumDict[name] >= 1000 {
            indexsumDict[name] += index + 1000
            if indexsumDict[name] < min {
                min = indexsumDict[name]
            }
        }
    }
    var ans []string
    for name := range indexsumDict {
        if indexsumDict[name] == min {
            ans = append(ans, name)
        }
    }
    return ans
}
